libnmstate/nm/lldp.py

```diff
@@ -0,0 +1,330 @@
+    3: LLDP.Neighbors.Port.TYPE_MAC,
+    4: LLDP.Neighbors.Port.TYPE_NETWORK_ADDRESS,
+    5: LLDP.Neighbors.Port.TYPE_INTERFACE_NAME,
+    6: LLDP.Neighbors.Port.TYPE_AGENT_CIRCUIT,
+    7: LLDP.Neighbors.Port.TYPE_LOCAL,
+}
+
+
+def apply_lldp_setting(con_setting, iface_desired_state):
+    lldp_status = iface_desired_state.get(LLDP.CONFIG_SUBTREE, {}).get(
+        LLDP.ENABLED, None
+    )
+    if lldp_status is not None:
+        lldp_status = int(lldp_status)
+        con_setting.setting.props.lldp = lldp_status
+
+
+def get_info(nm_client, nmdev):
+    """
+    Provides the current LLDP neighbors information
+    """
+    lldp_status = _get_lldp_status(nm_client, nmdev)
+    info = {}
+    if lldp_status == -1 or not lldp_status:
+        info[LLDP.ENABLED] = False
+    else:
+        info[LLDP.ENABLED] = True
+        _get_neighbors_info(info, nmdev)
+
+    return {LLDP.CONFIG_SUBTREE: info}
+
+
+def _get_lldp_status(nm_client, nmdev):
+    """
+    This is needed due to a missing property in NM.Device
+    Ref: https://bugzilla.redhat.com/1832273
+    """
+    lldp_status = None
+    con_profile = connection.ConnectionProfile(nm_client)
+    con_profile.import_by_device(nmdev)
+    if con_profile.profile:
+        con_setting = con_profile.profile.get_setting_connection()
+        if con_setting:
+            lldp_status = con_setting.get_lldp()
+
+    return lldp_status
+
+
+def _get_neighbors_info(info, nmdev):
+    neighbors = nmdev.get_lldp_neighbors()
+    info_neighbors = []
+    for neighbor in neighbors:
+        n_info = {}
+        _add_neighbor_system_info(neighbor, n_info)
+        _add_neighbor_chassis_info(neighbor, n_info)
+        _add_neighbor_port_info(neighbor, n_info)
+        _add_neighbor_bridge_info(neighbor, n_info)
+        _add_neighbor_vlans_info(neighbor, n_info)
+        _add_neighbor_macphy_info(neighbor, n_info)
+        _add_neighbor_port_vlans_info(neighbor, n_info)
+        _add_neighbor_management_addresses(neighbor, n_info)
+        info_neighbors.append(n_info)
+
+    if info_neighbors:
+        info[LLDP.NEIGHBORS_SUBTREE] = info_neighbors
+
+
+def _add_neighbor_system_info(neighbor, info):
+    sys_name = neighbor.get_attr_value(NM.LLDP_ATTR_SYSTEM_NAME)
+    if sys_name:
+        info[LLDP.Neighbors.SYSTEM_NAME] = sys_name.get_string()
+
+    sys_desc = neighbor.get_attr_value(NM.LLDP_ATTR_SYSTEM_DESCRIPTION)
+    if sys_desc:
+        info[LLDP.Neighbors.SYSTEM_DESCRIPTION] = sys_desc.get_string()
+
+    sys_caps = neighbor.get_attr_value(NM.LLDP_ATTR_SYSTEM_CAPABILITIES)
+    if sys_caps:
+        info[LLDP.Neighbors.SYSTEM_CAPABILITIES] = _decode_sys_caps(
+            sys_caps.get_uint32()
+        )
+
+
+def _decode_sys_caps(code):
+    capabilities = []
+    prefix_length = len("0b")
+    bits = bin(code)[prefix_length:]
+    for i in range(1, len(bits) + 1):
+        try:
+            if bits[-i] == "1":
+                capabilities.append(LLDP_CAP_NAMES[i - 1])
```

Comment:
In my opinion, it would be more robust and easier to understand if `LLDP_CAP_NAMES` contains the actual bitmasks like

```
LDDP_CAPABILITY_NAMES = {
0b1:     LLDP.Neighbors.SYSTEM_CAPABILITY_OTHER,
0b10:￼    LLDP.Neighbors.SYSTEM_CAPABILITY_REPEATER,
0b100:￼    LLDP.Neighbors.SYSTEM_CAPABILITY_MAC_BRIDGE,
0b1000:￼    LLDP.Neighbors.SYSTEM_CAPABILITY_WLAN_AC,
0b1_0000:￼    LLDP.Neighbors.SYSTEM_CAPABILITY_ROUTER,
...
}
```

and then use something like

```
capabilities = []
for mask, capability in LLDP_CAPABILITY_NAMES:
    if code & mask:
        capabilities.append(capability)
return capabilities
```

What do you think?

Reply:
Hm. That is a nice option. I agree with this. Thanks!

Reply:
Done! Thanks :-)